middleware/rate_limiter.go

```diff
@@ -0,0 +1,114 @@
+package middleware
+
+import (
+	"github.com/labstack/echo/v4"
+	"golang.org/x/time/rate"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// RateLimiterStore is the interface to be implemented by custom stores.
+type RateLimiterStore interface {
+	// Stores for the rate limiter have to implement the Allow method
+	Allow(identifier string) bool
+}
+
+type (
+	// RateLimiterConfig defines the configuration for the rate limiter
+	RateLimiterConfig struct {
+		Skipper    Skipper
+		BeforeFunc BeforeFunc
+		// SourceFunc uses echo.Context to extract the identifier for a visitor
+		SourceFunc func(context echo.Context) string
+		// Store defines a store for the rate limiter
+		Store        RateLimiterStore
+		ErrorHandler func(context echo.Context) error
+	}
+)
+
+// DefaultRateLimiterConfig defines default values for RateLimiterConfig
+var DefaultRateLimiterConfig = RateLimiterConfig{
+	Skipper: DefaultSkipper,
+	SourceFunc: func(ctx echo.Context) string {
+		return ctx.RealIP()
+	},
+	ErrorHandler: func(context echo.Context) error {
+		return context.JSON(http.StatusTooManyRequests, nil)
+	},
+}
+
+// RateLimiter returns a rate limiting middleware
```

Comment:
I wanted to make it default but since the rate and burst of the `RateLImiterMemoryStore` have to be specified by the user explicitly I figured it'd be best to not make it default.

maybe if there are standard industry default for rate and burst of a rate limiting service then I could set them and use it as a default instead.